src/compiler/determinacy_checker.ml

```diff
@@ -1,235 +1,762 @@
+      | Some AssumedFunctional -> Functional
+      | Some e ->
+          let modes = get_mode_func ~loc:t.loc n e in
+
+          Format.eprintf "The functionality of %a is %a@." F.pp n pp_functionality e;
+
+          Format.eprintf "Valid_call with functionality:%a, arg:[%a], mode:[%a]@." pp_functionality e
+            (pplist ScopedTerm.pretty ",") args (pplist pp_arg_mode ",") modes;
+
+          let valid_call = valid_call ctx e args modes in
+
+          Format.eprintf "Valid call for %a is %a@." F.pp n pp_functionality valid_call;
+
+          if valid_call <> Any then infer_outputs ctx e args modes
+          else (
+            infer_outputs_fail ctx e args modes |> ignore;
+            Any)
+    (* returns Any if the call is not valid *)
+    and valid_call ctx =
+      fold_on_modes
+        (fun e l r ->
+          let inferred = infer ctx e in
+          if functionality_leq ~loc:e.ScopedTerm.loc inferred l then r else Any)
+        (fun _ _ r -> r)
+    and infer_outputs ctx =
+      fold_on_modes
+        (fun _ _ r -> r)
+        (fun t l r ->
+          Format.eprintf "Inferring output %a with func %a@." ScopedTerm.pretty t pp_functionality l;
+          match t.ScopedTerm.it with
+          | Var (n, []) ->
+              add_env ~loc:t.loc n ~v:l;
+              r
+          | Var (n, args) ->
+              let v = get_funct_of_term ctx t |> Option.get in
+              add_env ~loc:t.loc n ~v;
+              r
+          | _ -> if functionality_leq ~loc:t.loc (infer ctx t) l then r else Any)
+    and infer_outputs_fail ctx =
+      fold_on_modes
+        (fun _ _ r -> r)
+        (fun t _ r ->
+          all_vars2any t;
+          r)
+    and infer ctx ({ it; loc } as t : ScopedTerm.t) : functionality =
+      match it with
+      | ScopedTerm.Const _ | CData _ -> get_funct_of_term ctx t |> Option.get
+      | Discard -> Functional
+      | Var (n, []) -> ( match Env.get !env n with None -> Any | Some e -> e)
+      | Var (n, args) -> infer_app n ctx t args
+      | Lam (None, _, _type, t) -> arr Any (infer ctx t)
+      | Lam (Some vname, _, _type, t) ->
+          let v = Compilation.TypeAssignment.type_ass_2func ~loc global _type in
+          Format.eprintf "Going under lambda %a with func: %a@." F.pp (fst vname) pp_functionality v;
+          arr Any (infer (add_ctx ~loc ctx vname ~v) t)
+      | Impl (true, _hd, t) -> infer ctx t (* TODO: hd is ignored *)
+      | Impl (false, _, t) -> infer ctx t (* TODO: this is ignored *)
+      | App (_, n, x, [ y ]) when F.equal F.eqf n || F.equal F.isf n || F.equal F.asf n ->
+          Format.eprintf "Calling inference for unification between \n - (@[%a@])\n - (%a)@." ScopedTerm.pretty x
+            ScopedTerm.pretty y;
+          let f1, f2 = (infer ctx x, infer ctx y) in
+          Format.eprintf "Inferred are \n - %a\n -%a@." pp_functionality f1 pp_functionality f2;
+          unify_func x y f1 f2;
+          Functional
+      | App (_, n, x, xs) when F.equal F.andf n ->
```

Comment:
It is probably better to add explicit And, Or... nodes in scoped term.
But in the meanwhile you have to check that the scope is Global.